### PR TITLE
Add integration test for multicluster headless svc

### PR DESCRIPTION
**Pull request by importhuman**
This is a first draft. It
- adds changes to bin/_test-helpers.sh to pass the headless services
flag and also to run the new test
- adds a new target for the integration test

See #6563

Signed-off-by: Ujjwal Goyal <importujjwal@gmail.com>

---
I've started off with setting up the integration test, thus the changes so far. I've based the new target manifest on [target1/emojivoto-no-bot.yml](https://github.com/linkerd/linkerd2/blob/main/test/integration/multicluster/target1/testdata/emojivoto-no-bot.yml). Is this appropriate? I tried making the `web` Deployment a StatefulSet with a corresponding HeadlessService, but was unable to access the application via port-forwarding (I didn't expect to, but wanted to check what happened). 

Thanks!

### CODE REVIEW

File: bin/_test-helpers.sh
Comment (importhuman) [reply bin/_test-helpers.sh:526]: Do we need `run_test "$test_directory/install_test.go" --multicluster --certs-path "$tmp"` again as well? As this is already being done in line 511 and hence in the k3d-target cluster?

**Comment by importhuman** [comment]
This commit doesn't add much, I've removed the line already present in `bin/_test-helpers.sh` and changed the target manifest [to this one that was linked above.](https://github.com/mateiidavid/l2d-k3d-statefulset/blob/main/west/nginx-statefulset.yml)
I'm trying to understand the flow of what we're doing here before adding the new test, please correct me wherever wrong:
- For tests involving multiclusters, we need atleast 2 clusters. This is broadly what's being done by [`run_multicluster_test()` in bin/_test-helpers.sh](https://github.com/linkerd/linkerd2/blob/main/bin/_test-helpers.sh#L504), along with running the tests. 
- `k3d-source` and `k3d-target` are our two contexts/clusters. 
- `target1` installs emojivoto in the target cluster (includes 3 deployments).
- `link` is what will connect our clusters.
- `source` runs the command `linkerd multicluster gateways`, for what I'm guessing is to check whether the clusters are connected (CLI describes it as "Display stats information about the gateways in target clusters"). Then, `source` installs the vote-bot in the source cluster (1 deployment).
- `target2` does what the comment in its file says: 
> TestTargetTraffic inspects the target cluster's web-svc pod to see if the source cluster's vote-bot has been able to hit it with requests.

Since we've already set up our clusters and workloads, there's no YAML manifest for `target2`.

Now for the new integration test to be added:
- In `nginx-statefulset.yml` (same as the one linked above), why do we need to specify `clusterIPs: None` after setting `clusterIP: None` just above?
- From the uses of slow cooker I've seen in the codebase, they all are from YAML manifests. 
> In the cluster that we create the mirrors in, it would make sense to use slow-cooker and just generate a light amount of load to the mirrored headless service.

We're creating mirrors in source cluster, so would adding slow-cooker to the manifest for source cluster affect the existing tests, or since we aren't using them in the existing ones, it is alright to modify the source YAML manifest? 

Please let me know if I've misinterpreted something else but haven't mentioned it.  Thanks in advance!
@kleimkuhler @mateiidavid

File: bin/_test-helpers.sh
Comment (mateiidavid) [reply bin/_test-helpers.sh]: Probably not (@kleimkuhler can correct me if I'm wrong here). Looking at the layout of the tests now, we first install everything in the `target` and `source` contexts -- looking at `target2` for example, we can see we skip creation (the test itself just tests traffic, so it makes sense).

I'd omit it, our clusters are created, we've tested installation already and traffic, third target can deal exclusively with _installing statefulsets_ or something similar.

File: test/integration/multicluster/target3/testdata/nginx-statefulset.yml
Comment (mateiidavid) [inline test/integration/multicluster/target3/testdata/nginx-statefulset.yml:27]: fwiw: one thing I'd like to test with headless services is scaling. Here's a suggestion, what if we first switch this replica count to 1 (or 2) -- with one replica, the test will start up a bit faster (StatefulSets start in order, unlike pods there's a bit of a dependency between replicas so it takes longer for them to come up) -- we can then scale up to 2 replicas and test that additional services have been mirrored in the source cluster. Luckily for us, StatefulSets have predictable names so we can easily assert for example an Endpoints object has the right state after scaling up.

Not a strict necessity, but it would definitely be good to test.

**Comment by mateiidavid** [comment]
Hey @importhuman, glad to see you're making some progress here! I'll answer some of the questions that you have to try and unblock you a little bit, integration tests can be a bit harder to think of at first since they require more context.

> This commit doesn't add much, I've removed the line already present in `bin/_test-helpers.sh` and changed the target manifest [to this one that was linked above.](https://github.com/mateiidavid/l2d-k3d-statefulset/blob/main/west/nginx-statefulset.yml)
> I'm trying to understand the flow of what we're doing here before adding the new test, please correct me wherever wrong:
> 
> * For tests involving multiclusters, we need atleast 2 clusters. This is broadly what's being done by [`run_multicluster_test()` in bin/_test-helpers.sh](https://github.com/linkerd/linkerd2/blob/main/bin/_test-helpers.sh#L504), along with running the tests.
> * `k3d-source` and `k3d-target` are our two contexts/clusters.
> * `target1` installs emojivoto in the target cluster (includes 3 deployments).
> * `link` is what will connect our clusters.
> * `source` runs the command `linkerd multicluster gateways`, for what I'm guessing is to check whether the clusters are connected (CLI describes it as "Display stats information about the gateways in target clusters"). Then, `source` installs the vote-bot in the source cluster (1 deployment).

Yeah, this is pretty much correct and how I understand the `run_multicluster_test()` function. The first thing that we do is set up the multicluster installation, that's a test in and of itself, we want to test that through the CLI tooling components start up properly, links can be established and so on. The `link` command is responsible for setting up a lot of the configuration needed to make multicluster work, that's why we check the gateways -- we want to make sure that the gateways are reachable and the networking components have been set up properly.

The second test (`target2`) seems to focus more on connectivity -- can we connect from one cluster to the other? We are fetching metrics there because it's the easiest way to test this. You'll see there's a `todo` to use `linkerd diagnostics --proxy`, that gives us the metrics from the proxy, certain metrics (such as `tcp_open_connections`) are a good indication that the application can either send or receive traffic (which is what we're looking to verify here).

> Now for the new integration test to be added:
> 
> * In `nginx-statefulset.yml` (same as the one linked above), why do we need to specify `clusterIPs: None` after setting `clusterIP: None` just above?

The short answer is, we don't. Applying it without `clusters` would work just as well. When it comes to Kubernetes primitives, a lot of the fields are not documented very thoroughly in the official docs, in such cases I'd go read the actual library code, there are a lot of informative comments (and it will always be the source of truth):

```go
// ClusterIPs is a list of IP addresses assigned to this service, and are
// usually assigned randomly.  If an address is specified manually, is
// in-range (as per system configuration), and is not in use, it will be
// allocated to the service; otherwise creation of the service will fail.
// This field may not be changed through updates unless the type field is
// also being changed to ExternalName (which requires this field to be
// empty) or the type field is being changed from ExternalName (in which
// case this field may optionally be specified, as describe above).  Valid
// values are "None", empty string (""), or a valid IP address.  Setting
// this to "None" makes a "headless service" (no virtual IP), which is
// useful when direct endpoint connections are preferred and proxying is
// not required.
// (...)
 // If this field is not specified, it will
// be initialized from the clusterIP field
```
[ref](https://pkg.go.dev/k8s.io/api/core/v1#ServiceSpec)

So, if we don't specify it, it will be taken from `clusterIP`.

> * From the uses of slow cooker I've seen in the codebase, they all are from YAML manifests.
> 
> We're creating mirrors in source cluster, so would adding slow-cooker to the manifest for source cluster affect the existing tests, or since we aren't using them in the existing ones, it is alright to modify the source YAML manifest?
> 

Well, my preference would be to keep the manifests separate and scoped for every test. I doubt they would interfere much, at the end of the day we are just deploying an application -- most of the tests (e.g `source_test`) will only check that the deployment that test is responsible for is up; e.g, we check `emojivoto` has spun up correctly. So, from this perspective, they wouldn't interfere, but it would be easier (and neat) to keep each manifest scoped to its test and the test's responsibilities.

Does this make sense?

File: test/integration/multicluster/target3/testdata/nginx-statefulset.yml
Comment (importhuman) [reply test/integration/multicluster/target3/testdata/nginx-statefulset.yml:27]: Not sure I understand, which scenario are you talking about here?
1. First write the integration test with 1 replica; after we do that we can try with 2 and that it still works.
2. In the same manifest, make sure that the 2nd replica starts setting up after the 1st is done. (If this is the one, don't tell me how to do it, I'd like to figure it out myself first)

Or is it something else?

**Comment by importhuman** [comment]
Thanks @mateiidavid! 

I'm still a bit unclear on the gateways part. Found this definition from [Google cloud docs:](https://cloud.google.com/kubernetes-engine/docs/how-to/deploying-gateways#internal-gateway)
> A Gateway resource represents a data plane that routes traffic in Kubernetes.

Without too much knowledge/detail, I'd say the gateways are involved in communication between clusters. So `link` sets up the required configuration *before resources in the source cluster are set up*, `source` first tests that the configuration is good to go before applying the source YAML manifest, and *then* `target2` actually checks if the clusters can connect. 
Would this be the basic process that is happening?

> It would be easier (and neat) to keep each manifest scoped to its test and the test's responsibilities.

Does this simply mean another manifest in `source/testdata` and apply it in `source_test.go`, or a manifest in `target3/testdata` and apply it in `_test-helpers.sh`?

Also regarding the slow-cooker manifest, would a deployment resource with the `buoyantio/slow_cooker` image be sufficient? Looking at the different uses of slow-cooker so far in the linkerd2 repo, it's either a job or deployment, and in some cases there is a service on the corresponding containerPort as well.

File: test/integration/multicluster/target3/testdata/nginx-statefulset.yml
Comment (mateiidavid) [reply test/integration/multicluster/target3/testdata/nginx-statefulset.yml:27]: Okay, for context, remember that a headless service that is mirrored (also known as a headless mirror) will be a "root" service. It will be a root because its endpoints object will contain addresses that belong to other services -- services that we create based on the pods in the StatefulSet (we call these endpoint mirrors).

What I'd like to test is the logic in the service-mirror but in a less controlled environment, i.e in the k3d clusters. We can apply the manifest with one replica, and that's part of our initial test ("can we deploy a StatefulSet with a headless service in a target cluster and see it mirrored in a source cluster?" is the first thing we should be testing here). A good follow-up thing to do would be to scale our StatefulSet up in the target cluster, rather than re-apply a manifest (with 2 replicas instead of 1) I'd like to just scale up the object itself (as an operator would in k8s through `kubectl` or whatever). We know ahead of time what the behaviour should be when a StatefulSet is scaled up (I'll leave these to you, there are a few, but for example we expect another `clusterIP` endpoint mirror to be created in the source cluster). So, this scaling test would just verify the assumptions that we have about the system in a real world environment. Hope this makes a bit more sense.

This is what I'd like to test, it's more in line with (1).

**Comment by mateiidavid** [comment]
The definition is good, maybe not simple enough. A gateway can be any application or server that intercepts connections and forwards them to where they have to go (also known as "sitting at the edge of the cluster", or "ingress"). Traffic has to flow in, but in order to allow traffic to a service in Kubernetes, a special kind of service is used (either a `NodePort` or a `Loadbalancer` type service). To make it easier to ingest traffic in the cluster, it's common for only one special kind of service to be used with a specific class of servers that just forward an incoming connection to where it would be supposed to go.

Our multicluster gateway makes use of a `LoadBalancer` service, but it's not a complex application. As a matter of fact, it's an empty container injected with the proxies. Whenever multicluster is installed, we create this "gateway" pod. When we link, we create a link resource that holds the address of the gateway's load balancer service (which is a publicly accessible IP address). The service mirror makes use of this address when a service is mirrored, the endpoints of the mirror service are set to the gateway's IP. Whenever a client makes a request to a mirror service, the request gets sent to the gateway where the proxy intercepts it and forwards it to where it was intended.

With that out of the way, yes, that is the basic process of what is happening. With target2, it's a bit more than just checking if we can connect, we're additionally also checking that the proxy routes the traffic to the service correctly. It's the whole scenario of sending a request from src --> target and making sure all involved systems behave accordingly. 

>

I'd create a manifest in `target3/testdata` and apply it in `target3/target_test.go`, look at `target/target_test.go` for an example:
```go
func TestInstallEmojivoto(t *testing.T) {
        if err := TestHelper.CreateDataPlaneNamespaceIfNotExists(context.Background(), "emojivoto", nil); err != nil {
                testutil.AnnotatedFatalf(t, "failed to create emojivoto namespace",
                        "failed to create emojivoto namespace: %s", err)
        }
        yaml, err := testutil.ReadFile("testdata/emojivoto-no-bot.yml")
        if err != nil {
                testutil.AnnotatedFatalf(t, "failed to read 'emojivoto-no-bot.yml'", "failed to read 'emojivoto-no-bot.yml': %s", err)
        }
        out, err := TestHelper.KubectlApply(yaml, "emojivoto")
        if err != nil {
                testutil.AnnotatedFatalf(t, "failed to install emojivoto", "failed to install emojivoto: %s\n%s", err, out)
        }
}
```

> Also regarding the slow-cooker manifest, would a deployment resource with the buoyantio/slow_cooker image be sufficient? Looking at the different uses of slow-cooker so far in the linkerd2 repo, it's either a job or deployment, and in some cases there is a service on the corresponding containerPort as well.

Well, this will be very dependent on how you choose to do things. Other tests may have other specific needs, each object has a responsibility in Kubernetes (or a function that it serves). Jobs run until completion, for example and services allow the pod to be discovered (so other clients can connect to it). Chances are we won't need either of them, we're not interested in the slow cooker at all here, we're interested in the nginx service and how that reacts to requests -- the slow cooker is just a convenient way for us to send those requests to the nginx service. I think a deployment should suffice.

**Comment by importhuman** [comment]
@mateiidavid Thanks, I'll try and do some more reading to understand the concept better.

> I'd create a manifest in `target3/testdata` and apply it in `target3/target_test.go`

~Yes that is what I'd thought, but don't we need the slow-cooker manifest to be applied in the source cluster?~

After writing the above line, I went back to #6563 to reread the issue description.
> We should test end-to-end how a headless service backed by a StatefulSet is created, exported and deleted (or unexported) and also test that an endpoint mirror belonging to the headless mirror can be used to send traffic to a specific instance from a StatefulSet. We can probably use the slow cooker as a client and nginx as the server.

Earlier I'd thought that as nginx is in `target`, slow-cooker needs to be in `source`, but now I'm guessing that is not the case. What we want is to set up nginx and slow-cooker (in target cluster), mirror the headless service (present in `target`) in the source cluster, then through the endpoint mirror, send traffic from src --> target. 
Would this be a somewhat accurate description?

**Comment by mateiidavid** [comment]
Yup, that's an accurate description @importhuman. However, we will only want to mirror the StatefulSet (nginx in our case), there is no point in mirroring the slow-cooker. Here's the set-up I imagined:
```
 +++++++++++++++        ++++++++++++++++++        ++++++++++++++++++++++++        ++++++++++++++++++
 + slow cooker + ====>  + nginx-mirror-1 +  ====> + gateway (tgt cluster)+ ====>  + nginx original +
 +++++++++++++++        ++++++++++++++++++        ++++++++++++++++++++++++        ++++++++++++++++++

where slow-cooker is deployed in source
and nginx is deployed in target

(1) ensure nginx can be exported as a StatefulSet
(2) ensure nginx can be scaled up/down in target cluster (and this is reflected in source)
(3) ensure slow cooker can reach nginx-mirror thro

**Comment by importhuman** [comment]
@mateiidavid is there a way I can check the resources during the integration test? For example, in the latest commit, I've applied the YAML manifest. Currently, I'm testing using the 'run test' option in VS Code, but that simply gives me the ok status. I tried doing `fmt.Println(out)`, but got no output. Further additions to the test will require me to check various resources, so am looking for help into that.

**Comment by kleimkuhler** [comment]
@importhuman I think the issue here is that you need to build the images first with `bin/docker-build`. Then, `image-load` should find those images and be able to load them into your test cluster.

**Comment by importhuman** [comment]
Finally resolved my issue of not being able to completely run `bin/docker-build`, so can resume work on this. On running
```
bin/tests --name multicluster `which linkerd`
```
I get a fail message when linkerd checks are run. The `actual` checks have the `pre-kubernetes-capability` section which the `expected` checks do not. Is this the cause, or should I provide the complete output of running the test? 

linkerd version:
Client version: stable-2.10.2
Server version: unavailable
```
Checking if there is a Kubernetes cluster available...[ok]
Checking if Linkerd resources exist on cluster...[ok]
Signature ok
subject=/CN=identity.linkerd.cluster.local
Getting CA Private Key
Test script: [install_test.go] Params: [--multicluster --certs-path /var/folders/<hiding path as it may contain key, unsure>]
--- FAIL: TestCheckPreInstall (3.03s)
    install_test.go:128: received unexpected output
        Expected:
        Linkerd core checks
        ===================

        kubernetes-api
        --------------
        √ can initialize the client
        √ can query the Kubernetes API

        kubernetes-version
        ------------------
        √ is running the minimum Kubernetes API version
        √ is running the minimum kubectl version

        pre-kubernetes-setup
        --------------------
        √ control plane namespace does not already exist
        √ can create non-namespaced resources
        √ can create ServiceAccounts
        √ can create Services
        √ can create Deployments
        √ can create CronJobs
        √ can create ConfigMaps
        √ can create Secrets
        √ can read Secrets
        √ can read extension-apiserver-authentication configmap
        √ no clock skew detected

        linkerd-version
        ---------------
        √ can determine the latest version
        √ cli is up-to-date

        Status check results are √

        Actual:
        kubernetes-api
        --------------
        √ can initialize the client
        √ can query the Kubernetes API

        kubernetes-version
        ------------------
        √ is running the minimum Kubernetes API version
        √ is running the minimum kubectl version

        pre-kubernetes-setup
        --------------------
        √ control plane namespace does not already exist
        √ can create non-namespaced resources
        √ can create ServiceAccounts
        √ can create Services
        √ can create Deployments
        √ can create CronJobs
        √ can create ConfigMaps
        √ can create Secrets
        √ can read Secrets
        √ can read extension-apiserver-authentication configmap
        √ no clock skew detected

        pre-kubernetes-capability
        -------------------------
        √ has NET_ADMIN capability
        √ has NET_RAW capability

        linkerd-version
        ---------------
        √ can determine the latest version
        √ cli is up-to-date

        Status check results are √
FAIL
FAIL	command-line-arguments	3.516s
FAIL
```

cc: @kleimkuhler @mateiidavid

**Comment by importhuman** [comment]
The commit applies a slow cooker deployment to the `source` cluster. How do I generate traffic? Tried adding a port 8080 to nginx-svc and referenced it in the slow-cooker deployment, but I'm pretty unclear on how to proceed. Thanks!

cc: @mateiidavid @kleimkuhler

File: test/integration/multicluster/target3/testdata/slow-cooker.yml
Comment (mateiidavid) [inline test/integration/multicluster/target3/testdata/slow-cooker.yml:24]: You can't use `http://nginx:8080` here, you'll have to use the fqn of an endpoint mirror. e.g `http://nginx-set-<no>-<target>.<namespace>.svc.cluster.local` where no is the instance number (e.g nginx-set-0) and target is the name of the target cluster.

**Comment by mateiidavid** [comment]
@importhuman I think what Alex is suggesting here is to use the secret as an _actual_ kubeconfig file that will allow you to connect to the target cluster. If you cannot connect to it (and e.g list pods and services) then it's likely something is wrong with the secret/kubeconfig file.

From `kubectl config`:
```
Modify kubeconfig files using subcommands like "kubectl config set
current-context my-context"

 The loading order follows these rules:

  1.  If the --kubeconfig flag is set, then only that file is loaded. The flag
may only be set once and no merging takes place.
  2.  If $KUBECONFIG environment variable is set, then it is used as a list of
paths (normal path delimiting rules for your system). These paths are merged.
When a value is modified, it is modified in the file that defines the stanza.
When a value is created, it is created in the first file that exists. If no
files in the chain exist, then it creates the last file in the list.
  3.  Otherwise, ${HOME}/.kube/config is used and no merging takes place.
  ```

You could, for example, do the following:
1. Get the secret, extract the kubeconfig file:
```sh
$ kubectl get secret cluster-credentials-target -o yaml > secret.yaml
$ vim secret.yaml # pull out kubeconfig into its own file, i.e target-config (no ext)
```
2. Use the kubeconfig file to see if you can interact with the target cluster:
```sh
$ kubectl get pods --kubeconfig target-config
```

**Comment by importhuman** [comment]
Hey @mateiidavid, here's what I've done after your reply, please let me know if there's a mistake in the steps.
In `source` cluster, in `linkerd-multicluster` namespace,
- Run `kubectl get secret cluster-credentials-target -o yaml` to get the secret. It has a "kubeconfig" field that is encoded. Decoded this by running `echo "<kubeconfig>" | base64 --decode > target-config`.
- `target-config` has the following decoded kubeconfig:

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <encoded-data>
    server: https://172.21.0.4:6443
  name: k3d-target
contexts:
- context:
    cluster: k3d-target
    namespace: linkerd-viz
    user: linkerd-service-mirror-remote-access-default
  name: k3d-target
current-context: k3d-target
kind: Config
preferences: {}
users:
- name: linkerd-service-mirror-remote-access-default
  user:
    token: <encoded-token>
```

- Run `kubectl get pods --kubeconfig target-config`.

After the last command, the terminal gets stuck with no output for a while, and then returns
`Unable to connect to the server: dial tcp 172.21.0.4:6443: i/o timeout`
I've tried running it with the `-A` flag as well, with the same result.

**Comment by mateiidavid** [comment]
@adleong @kleimkuhler @importhuman ready for a review :)

**Review by kleimkuhler — APPROVED** [review]
Looks good! Thanks for wrapping this up 👍

**Review by adleong — APPROVED** [review]

**Pull request merged**
